src/components/inputs/text/NameTokenSelector/NameTokenSelector.tsx

```diff
@@ -30,177 +44,512 @@ function NameTokenSelector({
     undefined,
   );
 
-  async function handlePDNTId(id: string) {
+  const listRef = useRef<HTMLDivElement>(null);
+  const [listPage, setListPage] = useState<number>(1);
+  const listItemCount = 3;
+
+  useEffect(() => {
+    selectedTokenCallback(
+      selectedToken ? new ArweaveTransactionID(selectedToken.id) : undefined,
+    );
+  }, [selectedToken]);
+
+  useEffect(() => {
+    if (!walletAddress) {
+      return;
+    }
+    getTokenList(walletAddress);
+  }, [walletAddress]);
+
+  useEffect(() => {
+    // token import update
+    if (searchText) {
+      handleTokenSearch(searchText);
+      setListPage(1);
+    }
+  }, [tokens]);
+
+  useEffect(() => {
+    if (!listRef.current) {
+      return;
+    }
+    document.addEventListener('click', handleClickOutside);
+
+    return () => {
+      document.removeEventListener('click', handleClickOutside);
+    };
+  }, [listRef]);
+
+  function handleClickOutside(e: any) {
+    e.preventDefault();
+    if (
+      listRef.current &&
+      e.target !== listRef.current &&
+      !listRef.current.contains(e.target)
+    ) {
+      setSearchText('');
+      setSearchActive(false);
+      setFilteredTokens(undefined);
+      setListPage(1);
+    }
+  }
+
+  async function getTokenList(
+    address: ArweaveTransactionID | undefined,
+    imports?: ArweaveTransactionID[],
+  ) {
+    try {
+      setLoading(true);
+      if (!address) {
+        throw new Error('No address provided');
+      }
+      const fetchedIds = await arweaveDataProvider.getContractsForWallet(
+        address,
+      );
+      if (!fetchedIds.ids) {
+        throw new Error(
+          'Unable to find any Name Tokens for the provided address',
+        );
+      }
+      if (imports) {
+        imports.map((id) =>
+          arweaveDataProvider
+            .validateTransactionTags({
+              id: id.toString(),
+              requiredTags: {
+                'App-Name': ['SmartWeaveContract'],
+              },
```

Comment:
we should likely load the state and validate the schema as well

e.g. getContractState() and then `new PDNTContract(state).isValid()`

Reply:
updated in https://github.com/ar-io/pdns-react/pull/229/commits/0b63d7bccbf324d95029f2a91e4a74e4e9116c0e